File: android/src/main/java/io/qonversion/capacitor/NoCodesPlugin.kt
package io.qonversion.capacitor

import android.app.Activity
import android.app.Application
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import io.qonversion.sandwich.ActivityProvider
import io.qonversion.sandwich.BridgeData
import io.qonversion.sandwich.NoCodesEventListener
import io.qonversion.sandwich.NoCodesPurchaseDelegateBridge
import io.qonversion.sandwich.NoCodesSandwich

private const val NoCodesEventName = "noCodesEvent"
private const val NoCodesPurchaseEventName = "noCodesPurchase"
private const val NoCodesRestoreEventName = "noCodesRestore"

@CapacitorPlugin(name = "NoCodes")
class NoCodesPlugin : Plugin() {
    private val noCodesSandwich by lazy {
        NoCodesSandwich()
    }

    private val noCodesEventListener = object : NoCodesEventListener {
        override fun onNoCodesEvent(event: NoCodesEventListener.Event, payload: BridgeData?) {
            val eventData = JSObject().apply {
                put("name", event.key)
                put("payload", payload?.toJSObject() ?: JSObject())
            }
            notifyListeners(NoCodesEventName, eventData)
        }
    }

    private val purchaseDelegateBridge = object : NoCodesPurchaseDelegateBridge {
        override fun purchase(product: BridgeData) {
            notifyListeners(NoCodesPurchaseEventName, product.toJSObject())
        }

        override fun restore() {
            notifyListeners(NoCodesRestoreEventName, JSObject())
        }
    }

    @PluginMethod
    fun initialize(call: PluginCall) {
        val context = context?.applicationContext
            ?: return call.reject("Can't get application context for NoCodes initialization", "InitializationError")
        val projectKey = call.getString("projectKey")
            ?: return call.noNecessaryDataError("projectKey")

        val proxyUrl = call.getString("proxyUrl")
        val locale = call.getString("locale")

        noCodesSandwich.initialize(context, projectKey, proxyUrl, locale = locale)
        noCodesSandwich.setDelegate(noCodesEventListener)

        val source = call.getString("source")
        val version = call.getString("version")
        if (source != null && version != null) {
            noCodesSandwich.storeSdkInfo(context, source, version)
        }

        call.resolve()
    }

    @PluginMethod
    fun setScreenPresentationConfig(call: PluginCall) {
        val configData = call.getObject("configData")?.toMap()
            ?: return call.noNecessaryDataError("configData")
        val contextKey = call.getString("contextKey")

        noCodesSandwich.setScreenPresentationConfig(configData, contextKey)
        call.resolve()
    }

    @PluginMethod
    fun showScreen(call: PluginCall) {
        val contextKey = call.getString("contextKey")
            ?: return call.noNecessaryDataError("contextKey")

        noCodesSandwich.showScreen(contextKey)
        call.resolve()
    }

    @PluginMethod
    fun close(call: PluginCall) {
        noCodesSandwich.close()
        call.resolve()
    }

    @PluginMethod
    fun setLocale(call: PluginCall) {
        val locale = call.getString("locale")
        noCodesSandwich.setLocale(locale)
        call.resolve()
    }

    @PluginMethod
    fun setPurchaseDelegate(call: PluginCall) {
        noCodesSandwich.setPurchaseDelegate(purchaseDelegateBridge)
        call.resolve()
    }

    @PluginMethod
    fun delegatedPurchaseCompleted(call: PluginCall) {
        noCodesSandwich.delegatedPurchaseCompleted()
        call.resolve()
    }

    @PluginMethod
    fun delegatedPurchaseFailed(call: PluginCall) {
        val errorMessage = call.getString("errorMessage")
            ?: return call.noNecessaryDataError("errorMessage")

        noCodesSandwich.delegatedPurchaseFailed(errorMessage)
        call.resolve()
    }

    @PluginMethod
    fun delegatedRestoreCompleted(call: PluginCall) {
        noCodesSandwich.delegatedRestoreCompleted()
        call.resolve()
    }

    @PluginMethod
    fun delegatedRestoreFailed(call: PluginCall) {
        val errorMessage = call.getString("errorMessage")
            ?: return call.noNecessaryDataError("errorMessage")

        noCodesSandwich.delegatedRestoreFailed(errorMessage)
        call.resolve()
    }
}

File: ios/Sources/QonversionPlugin/NoCodesPlugin.swift
import Foundation
import Capacitor
import QonversionSandwich

private let NoCodesEventName = "noCodesEvent"
private let NoCodesPurchaseEventName = "noCodesPurchase"
private let NoCodesRestoreEventName = "noCodesRestore"

@objc(NoCodesPlugin)
public class NoCodesPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NoCodesPlugin"
    public let jsName = "NoCodes"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setScreenPresentationConfig", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showScreen", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "close", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setPurchaseDelegate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setLocale", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "delegatedPurchaseCompleted", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "delegatedPurchaseFailed", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "delegatedRestoreCompleted", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "delegatedRestoreFailed", returnType: CAPPluginReturnPromise)
    ]

    var noCodesSandwich: NoCodesSandwich?
    var purchaseDelegateBridge: NoCodesPurchaseDelegateBridgeImpl?

    override public func load() {
        noCodesSandwich = NoCodesSandwich(noCodesEventListener: self)
    }

    @objc func initialize(_ call: CAPPluginCall) {
        guard let projectKey = call.getString("projectKey") else {
            return call.noNecessaryDataError()
        }

        let proxyUrl = call.getString("proxyUrl")
        let locale = call.getString("locale")

        noCodesSandwich?.initialize(projectKey: projectKey, proxyUrl: proxyUrl, locale: locale)

        if let source = call.getString("source"),
           let version = call.getString("version") {
            noCodesSandwich?.storeSdkInfo(source: source, version: version)
        }

        call.resolve()
    }

    @objc func setScreenPresentationConfig(_ call: CAPPluginCall) {
        guard let configData = call.getObject("configData") else {
            return call.noNecessaryDataError()
        }

        let contextKey = call.getString("contextKey")
        DispatchQueue.main.async { [weak self] in
            self?.noCodesSandwich?.setScreenPresentationConfig(configData, forContextKey: contextKey)
        }
        call.resolve()
    }

    @objc func showScreen(_ call: CAPPluginCall) {
        guard let contextKey = call.getString("contextKey") else {
            return call.noNecessaryDataError()
        }

        DispatchQueue.main.async { [weak self] in
            self?.noCodesSandwich?.showScreen(contextKey)
        }
        call.resolve()
    }

    @objc func close(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            self?.noCodesSandwich?.close()
        }
        call.resolve()
    }

    @objc func setLocale(_ call: CAPPluginCall) {
        let locale = call.getString("locale")
        noCodesSandwich?.setLocale(locale)
        call.resolve()
    }

    @objc func setPurchaseDelegate(_ call: CAPPluginCall) {
        purchaseDelegateBridge = NoCodesPurchaseDelegateBridgeImpl { [weak self] productData in
            self?.notifyListeners(NoCodesPurchaseEventName, data: productData)
        } onRestore: { [weak self] in
            self?.notifyListeners(NoCodesRestoreEventName, data: [:])
        }

        if let bridge = purchaseDelegateBridge {
            noCodesSandwich?.setPurchaseDelegate(bridge)
        }

        call.resolve()
    }

    @objc func delegatedPurchaseCompleted(_ call: CAPPluginCall) {
        noCodesSandwich?.delegatedPurchaseCompleted()
        call.resolve()
    }

    @objc func delegatedPurchaseFailed(_ call: CAPPluginCall) {
        guard let errorMessage = call.getString("errorMessage") else {
            return call.noNecessaryDataError()
        }

        noCodesSandwich?.delegatedPurchaseFailed(errorMessage)
        call.resolve()
    }

    @objc func delegatedRestoreCompleted(_ call: CAPPluginCall) {
        noCodesSandwich?.delegatedRestoreCompleted()
        call.resolve()
    }

    @objc func delegatedRestoreFailed(_ call: CAPPluginCall) {
        guard let errorMessage = call.getString("errorMessage") else {
            return call.noNecessaryDataError()
        }

        noCodesSandwich?.delegatedRestoreFailed(errorMessage)
        call.resolve()
    }
}

extension NoCodesPlugin: NoCodesEventListener {
    public func noCodesDidTrigger(event: String, payload: [String: Any]?) {
        let eventData: [String: Any] = [
            "name": event,
            "payload": payload ?? [:]
        ]
        self.notifyListeners(NoCodesEventName, data: eventData)
    }
}

class NoCodesPurchaseDelegateBridgeImpl: NoCodesPurchaseDelegateBridge {
    private let onPurchase: ([String: Any]) -> Void
    private let onRestore: () -> Void

    init(onPurchase: @escaping ([String: Any]) -> Void, onRestore: @escaping () -> Void) {
        self.onPurchase = onPurchase
        self.onRestore = onRestore
    }

    func purchase(_ product: [String: Any]) {
        onPurchase(product)
    }

    func restore() {
        onRestore()
    }
}
